### PR TITLE
added recursive list search for headings

### LINKED ISSUE #2308: Markdown heading level handling in docstring
Consider docstring:

```
"""
    foo(x, f, k)

Do stuff

# Arguments
- `x`: some argument
- `f`: some function
    ## Pattern
    - `f(t)`
    - `f(t, u)`
    - `f(t, u, c)`
- `k`: some other argument
"""
function foo(x, f, k)
end
```

In REPL help, the format is as expected: \"Pattern\" is indented and a lower level heading as \"Arguments\"

When making the documentation with Documenter on the other hand, \"Pattern\" gets translated to \<h2\> while \"Arguements\" only becomes \<strong\>, not \<h1\>. Therefore it looks quite off.

Using \<h1\> for \"Arguments\" is obviously not a good solution either
=> Idea: translate all heading levels in docstrings to \<strong\> instead of \<h_\>

### PR DESCRIPTION
**Pull request by 0815Creeper**
fixes #2308
added search through markdown-lists in docstrings to find headings for rewrite from heading to strong

### CODE REVIEW

File: src/expander_pipeline.jl [hunk 1]
```diff
@@ -961,12 +961,13 @@ function create_docsnode(docstrings, results, object, page, doc)
     slug = Documenter.slugify(object)
     anchor = Documenter.anchor_add!(doc.internal.docs, object, slug, page.build)
     docsnode = DocsNode(anchor, object, page)
-    # Convert docstring to MarkdownAST, convert Heading elements, and push to DocsNode
-    for (markdown, result) in zip(docstrings, results)
-        ast = convert(Node, markdown)
-        doc.user.highlightsig && highlightsig!(ast)
-        # The following 'for' corresponds to the old dropheaders() function
+    function recursive_heading_to_bold!(ast)
```
Comment (mortenpi) [inline src/expander_pipeline.jl:964]:
```suggestion
    function recursive_heading_to_bold!(ast::MarkdownAST.Node)
```

Comment (mortenpi) [reply src/expander_pipeline.jl:964]: Maybe let's pull this into a separate function, not a closure. It doesn't depend on any of the variables from `create_docsnode`, right?

**Review by mortenpi — COMMENTED** [review]
In addition to the comment, could you please:

1. Add a CHANGELOG.md note.
2. Add a test case. I think the best place for the test in this case would be one that shows up in https://documenter.juliadocs.org/stable/showcase/#Docstrings, so it should be somewhere in https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/DocumenterShowcase.jl

File: src/expander_pipeline.jl (hunk 1 above)
Comment (0815Creeper) [reply src/expander_pipeline.jl]: done

**Comment by 0815Creeper** [comment]
> In addition to the comment, could you please:
> 
>     1. Add a CHANGELOG.md note.
Done


> 
>     2. Add a test case. I think the best place for the test in this case would be one that shows up in https://documenter.juliadocs.org/stable/showcase/#Docstrings, so it should be somewhere in https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/DocumenterShowcase.jl
Not quite shure if I get what you said, but I tried

File: src/expander_pipeline.jl [hunk 2]
```diff
@@ -990,3 +981,21 @@ function highlightsig!(node::Node)
...
+        end
+        node.element isa MarkdownAST.Heading || continue
+        # node is a headingnode
+        boldnode = Node(MarkdownAST.Strong())
+        for textnode in collect(node.children)
+            push!(boldnode.children, textnode)
+        end
+        node.element = MarkdownAST.Paragraph()
+        push!(node.children, boldnode)
+    end
```
Comment (odow) [inline src/expander_pipeline.jl:1000]:
```suggestion
    end
    return
```

File: src/expander_pipeline.jl [hunk 3]
```diff
@@ -990,3 +981,21 @@ function highlightsig!(node::Node)
...
+end
```
Comment (odow) [inline src/expander_pipeline.jl:1001]:
```suggestion
end

```

File: src/expander_pipeline.jl [hunk 4]
```diff
@@ -990,3 +981,21 @@ function highlightsig!(node::Node)
...
 end
+
+function recursive_heading_to_bold!(ast::MarkdownAST.Node)
+    for node in ast.children
+        if node.element isa MarkdownAST.List
+            for child in node.children
+                recursive_heading_to_bold!(child)
+            end
+        end
+        node.element isa MarkdownAST.Heading || continue
```
Comment (odow) [inline src/expander_pipeline.jl:992]:
```suggestion
        if !(node.element isa MarkdownAST.Heading)
            continue
        end
```

File: docs/DocumenterShowcase.jl [hunk 5]
```diff
@@ -8,6 +8,8 @@ This is the [`DocumenterShowcase`](@ref), which contains
 
     This showcase page is included in Documenter version 0.24.0 and above.
 
+Headings that are part of docstrings are not rendered as headings but rather as bold text. Just like:
```
Comment (mortenpi) [inline docs/DocumenterShowcase.jl:11]:
```suggestion
Headings that are part of docstrings are not rendered as headings but rather as bold text:
```

File: docs/src/showcase.md [hunk 6]
```diff
@@ -324,6 +324,12 @@ We then see the same docstring as above
 DocumenterShowcase.bar
 ```
 
+We can also use nested lists and headings in docstrings
+
+```@docs
+DocumenterShowcase.baz
```
Comment (mortenpi) [inline docs/src/showcase.md:330]:
```suggestion
DocumenterShowcase.baz
```

File: docs/DocumenterShowcase.jl [hunk 7]
```diff
@@ -63,6 +65,26 @@ Docstring for `bar(::AbstractString)`.
 """
 bar(::AbstractString) = nothing
 
+"""
+    baz(x, f, k)
+
+Docstring for a more complex function.
```
Comment (mortenpi) [inline docs/DocumenterShowcase.jl:71]:
```suggestion
Function with a more complex docstring.
```

**Review by mortenpi — APPROVED** [review]
Did some cosmetic changes to the docs, but PR LGTM. Thank you @0815Creeper!

**Pull request merged**